src/sage/rings/function_field/drinfeld_modules/homset.py

```diff
@@ -421,3 +425,264 @@ def _element_constructor_(self, *args, **kwds):
+        # Reconstruct the Ore polynomial from the coefficients
+        basis = []
+        tau = domain.ore_polring().gen()
+        for basis_elem in sol:
+            basis.append(self(sum([sum([K_basis[j]*basis_elem[i*n + j]
+                               for j in range(n)])*(tau**i)
+                               for i in range(d + 1)])))
+        return basis
+
+    def basis(self):
+        r"""
+        Return a basis for the `\mathbb{F}_q[\tau^n]`-module of morphisms from
+        the domain to the codomain.
+
+        OUTPUT: a list of Drinfeld module morphisms.
+
+        EXAMPLES::
+
+            sage: Fq = GF(2)
+            sage: A.<T> = Fq[]
+            sage: K.<z> = Fq.extension(3)
+            sage: psi = DrinfeldModule(A, [z, z + 1, z^2 + z + 1])
+            sage: phi = DrinfeldModule(A, [z, z^2 + z + 1, z^2 + z])
+            sage: H = Hom(phi, psi)
+            sage: basis = H.basis()
+            sage: [b.ore_polynomial() for b in basis]
+            [(z^2 + 1)*t^2 + t + z + 1, (z^2 + 1)*t^5 + (z + 1)*t^4 + z*t^3 + t^2 + (z^2 + z)*t + z, z^2]
+
+            ::
+
+            sage: Fq = GF(5)
+            sage: A.<T> = Fq[]
+            sage: K.<z> = Fq.extension(3)
+            sage: phi = DrinfeldModule(A, [z, 3*z, 4*z])
+            sage: chi = DrinfeldModule(A, [z, 2*z^2 + 3, 4*z^2 + 4*z])
+            sage: H = Hom(phi, chi)
+            sage: H.basis()
+            []
+
+        ALGORITHM:
+
+            We return the basis of the kernel of a matrix derived from the
+            constraint that `\iota \phi_T = \psi_T \iota` for any morphism
+            `iota`. 
+        """
+        domain, codomain = self.domain(), self.codomain()
+        Fq = domain._Fq
+        K = domain.base_over_constants_field()
+        q = Fq.cardinality()
+        char = Fq.characteristic()
+        r = domain.rank()
+        n = K.degree(Fq)
+        qorder = logb(q, char)
+        K_basis = [K.gen()**i for i in range(n)]
+        dom_coeffs = domain.coefficients(sparse=False)
+        cod_coeffs = codomain.coefficients(sparse=False)
+        # The commutative polynomial ring in tau^n.
+        poly_taun = PolynomialRing(Fq, 'taun')
+
+        sys = Matrix(poly_taun, n**2, n**2)
+
+        # Build a linear system over the commutative polynomial ring
+        # Fq[tau^n]. The kernel of this system consists of all
+        # morphisms from domain -> codomain.
+        for j in range(n):
+            for k in range(n):
+                for i in range(r+1):
+                    # Coefficients of tau^{i + k} coming from the
+                    # relation defining morphisms of Drinfeld modules
+                    # These are elements of K, expanded in terms of
+                    # K_basis.
+                    c_tik = (dom_coeffs[i].frobenius(qorder*k)*K_basis[j] \
+                            - cod_coeffs[i]*K_basis[j].frobenius(qorder*i)) \
+                            .polynomial().coefficients(sparse=False)
```

Comment:
> I'm a bit surprised that this works, or that it does what you think it does. My understanding is the following:
> 
> * You aim at working over `Fq` (as a basis of `K`), and because `K` has degree `n` over `Fq`, `c_tik` should at the end have `n` elements.
> * It looks like the method `polynomial` returns the polynomial whose coefficients are that on the input _on the prime field_, i.e. Fp, in which case, its degree may be strictly larger than `n`.
> 
> In the below example, the polynomial has degree 3 while n equals 2:
> 
> ```
> sage: Fq = GF(3^2)
> sage: A.<T> = Fq[]
> sage: K.<z> = Fq.extension(2)
> sage: x = K.random_element()
> sage: x.polynomial?
> Docstring:     
>    Return self viewed as a polynomial over
>    "self.parent().prime_subfield()".
> 
>    EXAMPLES:
> 
>       sage: k.<b> = GF(5^2); k
>       Finite Field in b of size 5^2
>       sage: f = (b^2+1).polynomial(); f
>       b + 4
>       sage: type(f)
>       <class 'sage.rings.polynomial.polynomial_zmod_flint.Polynomial_zmod_flint'>
>       sage: parent(f)
>       Univariate Polynomial Ring in b over Finite Field of size 5
> Init docstring: Initialize self.  See help(type(self)) for accurate signature.
> File:           ~/sage/src/sage/rings/finite_rings/element_givaro.pyx
> Type:           builtin_function_or_method
> sage: x
> 2*z^2 + 2*z + 1
> sage: x.polynomial().coefficients(sparse=False)
> [1, 2, 2]
> ```
> 
> And this is more like your code:
> 
> ```
> sage: from sage.functions.log import logb
> sage: q = 4
> sage: Fq = GF(q)
> sage: A.<T> = Fq[]
> sage: _K_prime = Fq.extension(2)
> sage: _K.<z> = _K_prime.extension(3)
> sage: dom = DrinfeldModule(A, [z, 0, 1])
> sage: cod = DrinfeldModule(A, [z, 1, 1])
> sage: dom_coeffs = dom.gen().coefficients(sparse=False)
> sage: cod_coeffs = cod.gen().coefficients(sparse=False)
> sage: K = dom.base_over_constants_field()
> sage: n = K.degree(Fq)  # 6
> sage: qorder = logb(q, Fq.characteristic())
> sage: K_basis = [K.gen()**i for i in range(n)]
> sage: 
> sage: (i, j, k) = (2, 3, 4)
> sage: c_tik = (dom_coeffs[i].frobenius(qorder*k)*K_basis[j] \
> sage:          - cod_coeffs[i]*K_basis[j].frobenius(qorder*i)
> sage:         ).polynomial().coefficients(sparse=False)
> sage: c_tik
> [1, 1, 0, 0, 0, 1, 1, 0, 1, 1, 0, 1]
> ```
> 
> At the end, `c_tik` has length 12, while `n` equals 6. So while your method seems to output correct results, it may be working over Fp. This not at all a problem, but I believe the code and its comments should reflect this situation. Also, if I am not mistaken, then the following line (606) is never entered in.

Perhaps this is my fault for not marking this PR as a draft. At some point I realized that Fq.extension(n) returns a field constructed as a flat extension over $\mathbb{F}_p$ and not as a direct extension of $\mathbb{F}_q$, which made some of the computations I wanted to perform a bit more awkward to implement. As you may have noticed, all the tests are currently over the prime field case only, and this code will indeed fail for non-prime $q$. My plan was to research the finite field implementation a bit more to see what the most elegant way around this would be, ideally to have this work both for the flat extension setting and possibly for fields constructed as quotients of $\mathbb{F}_q[x]$. With my thesis largely done with I'll likely have more time to do this soon.

Reply:
> Perhaps this is my fault for not marking this PR as a draft.

Don't worry.

> With my thesis largely done with I'll likely have more time to do this soon.

Great. Indeed, it seems like `basis` kinda works, while `Fq_basis` simply does not (or not in all cases):

```
sage: Fq = GF(4)
sage: A.<T> = Fq[]
sage: K.<z> = Fq.extension(2)
sage: phi = DrinfeldModule(A, [z, z^2 + z + 1, 0, z^3 + z^2 + z + 1])
sage: psi = DrinfeldModule(A, [z, z^2 + 1, 1, z^3 + z^2 + z + 1])
sage: hom = Hom(phi, psi)
sage: hom.basis()
[Drinfeld Module morphism:
   From: Drinfeld module defined by T |--> (z^3 + z^2 + z + 1)*t^3 + (z^2 + z + 1)*t + z
   To:   Drinfeld module defined by T |--> (z^3 + z^2 + z + 1)*t^3 + t^2 + (z^2 + 1)*t + z
   Defn: t^2 + z*t + 1]
sage: hom.Fq_basis(2)
[]
```

In any case, any limitation to the prime field should be specified in the docs.

Reply:
> In any case, any limitation to the prime field should be specified in the docs.

I intend to remove this limitation, and there are several ways to do this (such, for example, doing linear algebra over $\mathbb{F}_p$ or using an explicit field embedding), though this raised the question of whether the implementation of Drinfeld modules should be able to handle fields constructed as explicit towers. For example, consider the following:

```
sage: Fq = GF(4)
sage: A.<T> = Fq[]
sage: ip = A.irreducible_element(2)
sage: K.<z> = Fq.extension(ip)
```
Currently, `K` as constructed above can't be used to create a Drinfeld module using our implementation, and I had planned to ask what opinions were on adding support for this.

Reply:
Hmm, thanks for pointing this out. I was not aware of this limitation; I think there is no reason to prevent the user from using this construction, and would definitely support the addition of this feature...

Reply:
> Hmm, thanks for pointing this out. I was not aware of this limitation; I think there is no reason to prevent the user from using this construction, and would definitely support the addition of this feature...

Ok I'll have a look into it then.

Reply:
Thank you. I'm happy to help (I won't have much time before summer tho, I'm focused on writing my thesis, and I will be away in July).

Reply:
That's fine, I don't think it's necessary going to be too difficult. The easiest approach would probably be to extend or add certain methods to work for quotient rings (such as `in_base` and `frobenius`).